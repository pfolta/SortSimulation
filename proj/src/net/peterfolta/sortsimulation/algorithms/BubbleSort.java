/*
 * SortSimulation - A visual representation of sorting algorithms
 * Copyright (C) 2008-2015 Peter Folta. All rights reserved.
 * 
 * Project:			SortSimulation 
 * Version:			2.0.1
 * Website:			http://www.peterfolta.net/software/sortsimulation
 * 
 * File:			Bubblesort.java
 * Created:			2008/11/29
 * Last modified:	2015/2/16
 * Author:			Peter Folta <mail@peterfolta.net>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.peterfolta.sortsimulation.algorithms;

import net.peterfolta.sortsimulation.common.ArrayTools;
import net.peterfolta.sortsimulation.common.interfaces.Sortable;
import net.peterfolta.sortsimulation.main.Main;

public class BubbleSort implements Sortable {
	
	private boolean interrupted = false;
	
	public void sort(int[] a, final int index) {
		for (int i = a.length-1; i >= 0 && !interrupted; i--) {
			for (int j = 0; j <= i-1 && !interrupted; j++) {
				if (a[j] > a[j+1]) {
					ArrayTools.swap(a, j, j+1);
				}
				
				try {
					Thread.sleep(Main.settings.getDelay().getDelay());
				} catch (InterruptedException exception) {
					interrupted = true;
					break;
				}
				
				Main.getGUI().getDisplay().asyncExec(new Runnable() {
					public void run() {
						Main.getGUI().getMainWindow().repaintCanvas(index);
					}
				});
			}
		}
	}

}