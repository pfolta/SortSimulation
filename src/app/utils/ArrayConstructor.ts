import "@/app/utils/Array";

declare global {
    interface ArrayConstructor {
        almostSorted: (size: number, distance: number) => number[];
        fewUnique: (size: number, uniqueCount: number) => number[];
        random: (size: number) => number[];
        range: <T>(start: number | T[], end?: number) => number[];
        reversed: (size: number) => number[];
        sorted: (size: number) => number[];
    }
}

/**
 * Create an array where every element is at most `distance` places away from its sorted position
 */
Array.almostSorted = (size: number, distance: number): number[] =>
    Array.sorted(size)
        .chunked(distance + 1)
        .flatMap((chunk) => chunk.shuffled());

/**
 * Create a randomly shuffled array containing `uniqueCount` unique numbers from 1 to `size`, e.g. [7, 3, 8, 8, 7, 4, 3, 1, 3, 4]
 */
Array.fewUnique = (size: number, uniqueCount: number): number[] => {
    const uniques = Array.random(size).slice(0, uniqueCount);
    return Array.range(size).map(() => uniques[Math.floor(Math.random() * uniques.length)]);
};

/**
 * Create a randomly shuffled array containing all numbers from 1 to `size`, e.g. [3, 5, 1, 2, 4]
 */
Array.random = (size: number): number[] => Array.sorted(size).shuffled();

/**
 * Create a range from `start` to less than `end` (inclusive of `start` but not `end`): [start, end)
 * If only one argument is passed, the range will start at 0: [0, end)
 * If an array is passed, the range will contain the indices of the array provided: [0, array.length)
 */
Array.range = <T>(start: number | T[], end?: number): number[] => {
    if (Array.isArray(start)) {
        return [...start.keys()];
    }

    if (end !== undefined) {
        return Array.from({ length: end - start }, (_, i) => start + i);
    }

    return [...Array(start).keys()];
};

/**
 * Create an array from `size` down to 1 in descending order, e.g. [5, 4, 3, 2, 1]
 */
Array.reversed = (size: number): number[] => Array.sorted(size).reverse();

/**
 * Create a sorted array from 1 to `size`, e.g. [1, 2, 3, 4, 5]
 */
Array.sorted = (size: number): number[] => Array.range(1, size + 1);

export {};
