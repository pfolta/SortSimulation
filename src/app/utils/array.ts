const isSorted = (array: number[]): boolean => {
    for (let i = 1; i < array.length; i++) {
        if (array[i - 1] > array[i]) {
            return false;
        }
    }

    return true;
};

export { isSorted };
