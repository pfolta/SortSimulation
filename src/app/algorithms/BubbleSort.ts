import { SortingAlgorithm } from "@/app/algorithms";
import { addToTrace, compare, compareGroup, createTrace, sortedGroup, swap, swapGroup, Trace } from "@/app/trace";
import "@/app/utils/Array";
import "@/app/utils/ArrayConstructor";

const BubbleSort: SortingAlgorithm = {
    id: "bubble_sort",

    sort: (array: number[]): Trace => {
        const trace = createTrace(array);

        let swapped = false;
        let i = array.length - 1;

        do {
            swapped = false;
            const sorted = sortedGroup(Array.range(i + 1, array.length));

            for (let j = 0; j < i; j++) {
                addToTrace(trace, array, compare, [compareGroup([j, j + 1]), sorted]);

                if (array[j] > array[j + 1]) {
                    array.swap(j, j + 1);
                    swapped = true;

                    addToTrace(trace, array, swap, [swapGroup([j, j + 1]), sorted]);
                }
            }

            i = i - 1;
        } while (swapped);

        addToTrace(trace, array, undefined, [sortedGroup(Array.range(array))]);

        return trace;
    }
};

export default BubbleSort;
