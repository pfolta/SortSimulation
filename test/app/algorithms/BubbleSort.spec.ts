import BubbleSort from "@/app/algorithms/BubbleSort";
import "@/app/utils/Array";

describe("Bubble Sort", () => {
    describe("sort", () => {
        const array = [6, 5, 10, 9, 8, 4, 7, 3, 1, 2];

        it("returns a trace starting with the initial array", () => {
            const trace = BubbleSort.sort([...array]);
            expect(trace[0].array).toEqual(array);
        });

        it("returns a trace containing the correct intermediate array steps", () => {
            const trace = BubbleSort.sort([...array]);
            expect(trace.map((step) => step.array)).toMatchSnapshot();
        });

        it("returns a trace containing the correct phases", () => {
            const trace = BubbleSort.sort([...array]);
            expect(trace.map((step) => step.phase?.key)).toMatchSnapshot();
        });

        it("returns a trace ending with the sorted array", () => {
            const trace = BubbleSort.sort([...array]);
            expect(trace[trace.length - 1].array.isSorted()).toBe(true);
        });
    });
});
