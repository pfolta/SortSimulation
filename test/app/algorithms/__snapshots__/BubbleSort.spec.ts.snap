// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bubble sort sort returns a trace containing the correct intermediate array steps 1`] = `"[[6,5,10,9,8,4,7,3,1,2],[6,5,10,9,8,4,7,3,1,2],[5,6,10,9,8,4,7,3,1,2],[5,6,10,9,8,4,7,3,1,2],[5,6,10,9,8,4,7,3,1,2],[5,6,9,10,8,4,7,3,1,2],[5,6,9,10,8,4,7,3,1,2],[5,6,9,8,10,4,7,3,1,2],[5,6,9,8,10,4,7,3,1,2],[5,6,9,8,4,10,7,3,1,2],[5,6,9,8,4,10,7,3,1,2],[5,6,9,8,4,7,10,3,1,2],[5,6,9,8,4,7,10,3,1,2],[5,6,9,8,4,7,3,10,1,2],[5,6,9,8,4,7,3,10,1,2],[5,6,9,8,4,7,3,1,10,2],[5,6,9,8,4,7,3,1,10,2],[5,6,9,8,4,7,3,1,2,10],[5,6,9,8,4,7,3,1,2,10],[5,6,9,8,4,7,3,1,2,10],[5,6,9,8,4,7,3,1,2,10],[5,6,8,9,4,7,3,1,2,10],[5,6,8,9,4,7,3,1,2,10],[5,6,8,4,9,7,3,1,2,10],[5,6,8,4,9,7,3,1,2,10],[5,6,8,4,7,9,3,1,2,10],[5,6,8,4,7,9,3,1,2,10],[5,6,8,4,7,3,9,1,2,10],[5,6,8,4,7,3,9,1,2,10],[5,6,8,4,7,3,1,9,2,10],[5,6,8,4,7,3,1,9,2,10],[5,6,8,4,7,3,1,2,9,10],[5,6,8,4,7,3,1,2,9,10],[5,6,8,4,7,3,1,2,9,10],[5,6,8,4,7,3,1,2,9,10],[5,6,4,8,7,3,1,2,9,10],[5,6,4,8,7,3,1,2,9,10],[5,6,4,7,8,3,1,2,9,10],[5,6,4,7,8,3,1,2,9,10],[5,6,4,7,3,8,1,2,9,10],[5,6,4,7,3,8,1,2,9,10],[5,6,4,7,3,1,8,2,9,10],[5,6,4,7,3,1,8,2,9,10],[5,6,4,7,3,1,2,8,9,10],[5,6,4,7,3,1,2,8,9,10],[5,6,4,7,3,1,2,8,9,10],[5,4,6,7,3,1,2,8,9,10],[5,4,6,7,3,1,2,8,9,10],[5,4,6,7,3,1,2,8,9,10],[5,4,6,3,7,1,2,8,9,10],[5,4,6,3,7,1,2,8,9,10],[5,4,6,3,1,7,2,8,9,10],[5,4,6,3,1,7,2,8,9,10],[5,4,6,3,1,2,7,8,9,10],[5,4,6,3,1,2,7,8,9,10],[4,5,6,3,1,2,7,8,9,10],[4,5,6,3,1,2,7,8,9,10],[4,5,6,3,1,2,7,8,9,10],[4,5,3,6,1,2,7,8,9,10],[4,5,3,6,1,2,7,8,9,10],[4,5,3,1,6,2,7,8,9,10],[4,5,3,1,6,2,7,8,9,10],[4,5,3,1,2,6,7,8,9,10],[4,5,3,1,2,6,7,8,9,10],[4,5,3,1,2,6,7,8,9,10],[4,3,5,1,2,6,7,8,9,10],[4,3,5,1,2,6,7,8,9,10],[4,3,1,5,2,6,7,8,9,10],[4,3,1,5,2,6,7,8,9,10],[4,3,1,2,5,6,7,8,9,10],[4,3,1,2,5,6,7,8,9,10],[3,4,1,2,5,6,7,8,9,10],[3,4,1,2,5,6,7,8,9,10],[3,1,4,2,5,6,7,8,9,10],[3,1,4,2,5,6,7,8,9,10],[3,1,2,4,5,6,7,8,9,10],[3,1,2,4,5,6,7,8,9,10],[1,3,2,4,5,6,7,8,9,10],[1,3,2,4,5,6,7,8,9,10],[1,2,3,4,5,6,7,8,9,10],[1,2,3,4,5,6,7,8,9,10],[1,2,3,4,5,6,7,8,9,10]]"`;
